File: C++/check-if-it-is-a-straight-line.cpp
class Solution {
public:
    bool checkStraightLine(vector<vector<int>>& coordinates) {
        
        int x1 = coordinates[0][0];
        int y1 = coordinates[0][1];

        int x2 = coordinates[1][0];
        int y2 = coordinates[1][1];
        
        double slope1 = (double)(x1 - x2) / (y1 - y2);

        int sz = coordinates.size();
        
        for (int i = 2; i < sz; i++) {
            x1 = x2;
            y1 = y2;
            x2 = coordinates[i][0];
            y2 = coordinates[i][1];
            
            double slope2 = (double)(x1 - x2) / (y1 - y2);

            if ((y1 != y2) && (fabs(slope1 - slope2) > 0.001)) {
                return false;
            }
            
            slope1 = slope2;
        }
        
        return true;
    }
};

File: TypeScript/check-if-it-is-a-straight-line.ts
function checkStraightLine(coordinates: number[][]): boolean {
    const n = coordinates.length
    const dx = coordinates[0][0] - coordinates[1][0]
    const dy = coordinates[0][1] - coordinates[1][1]
    if (dy === 0) { // horizontal line
        for (let i = 1; i < n; i++) {
            const dy2 = coordinates[0][1] - coordinates[i][1]
            if (dy2 !== 0) {
                return false
            }
        }        
    } else {
        const slope = dx / dy
        for (let i = 1; i < n; i++) {
            const dx2 = coordinates[0][0] - coordinates[i][0]
            const dy2 = coordinates[0][1] - coordinates[i][1]
            const slope2 = dx2 / dy2
            if (dy2 === 0 || Math.abs(slope - slope2) > 0.0000001) { // floating-point tolerance
                return false
            }
        }        
    }
    return true
};